frontend/catalyst/jax_primitives.py

```diff
@@ -368,6 +403,133 @@ def _print_lowering(jax_ctx: mlir.LoweringRuleContext, *args, string=None, memre
+    for i in range(len(module.body.operations)):
+        if module.body.operations[i].operation.name == "transform.named_sequence":
+            return module.body.operations[i].operation.results
+
+    functype = ir.FunctionType.get(inputs=[transform_func_type], results=[])
+    functype_attr = ir.TypeAttr.get(functype)
+
+    # add the transform.with_named_sequence unit attr to the parent module
+    with_named_sequence_attr = ir.UnitAttr.get(jax_ctx.module_context.context)
+    module.operation.attributes["transform.with_named_sequence"] = with_named_sequence_attr
+
+    # Insert the transform.named_sequence op into the module to be transformed
+    # Note that InsertionPoint(Block) inserts after the last operation but still inside the block.
+    # Hence we insert at the end, inside the module
+    with ir.InsertionPoint(module.body):
+        named_sequence_op = NamedSequenceOp(
+            sym_name="__transform_main",
+            function_type=functype_attr,
+        )
+
+        # transform.named_sequence op is the "main function" of the transform dialect
+        # and thus needs an entry block (which also should be its only block)
+        # The argument of the block is the target function to run the transform sequence on
+        bb = ir.Block.create_at_start(named_sequence_op.body, arg_types=[transform_func_type])
+
+        # The transform.named_sequence needs a terminator called "transform.yield"
+        with ir.InsertionPoint(bb):
+            transform_yield_op = TransformYieldOp(operands_=[])  # pylint: disable=unused-variable
+
+    return named_sequence_op.results
+
+
+#
+# apply_registered_pass
+#
+@apply_registered_pass_p.def_abstract_eval
+def _apply_registered_pass_abstract_eval(*args, pass_name, options=None):
+    return AbstractTransformFunc()
+
+
+@apply_registered_pass_p.def_impl
+def _apply_registered_pass_def_impl(*args, pass_name, options=None):  # pragma: no cover
+    raise NotImplementedError()
+
+
+def _apply_registered_pass_lowering(
+    jax_ctx: mlir.LoweringRuleContext, *args, pass_name, options=None
+):
+    module = jax_ctx.module_context.module
+    named_sequence_op = None
+    for i in reversed(range(len(module.body.operations))):
+        # transform.named_sequence usually is at the end of the module, so look for it from the end
+        if module.body.operations[i].operation.name == "transform.named_sequence":
+            named_sequence_op = module.body.operations[i].operation
+            break
+    if named_sequence_op is None:
+        raise RuntimeError(
+            "transform.apply_registered_pass must be placed in a transform.named_sequence!"
+        )
+
+    # If there already is a apply_registered_pass, insert after the existing one
+    # Note that ir.InsertionPoint(op) sets the insertion point to immediately BEFORE the op
+    named_sequence_op_block = named_sequence_op.regions[0].blocks[0]
+    first_op_in_block = named_sequence_op_block.operations[0].operation
```

Comment:
Shouldn't new passes always be added before the yield?

Reply:
Here, an interesting test would be to create passes (which do not have an MLIR implementation yet) and just stack them in the same function and lit-test the order of the passes.

Reply:
The comment is what made me ask the question.

>     # If there already is a apply_registered_pass, insert after the existing one

To me this reads as insert after the first one. If there are two or more, then it will still insert after the first one.

Reply:
We will always insert after the last existing pass.

Reply:
> Here, an interesting test would be to create passes (which do not have an MLIR implementation yet) and just stack them in the same function and lit-test the order of the passes.

The `-transform-interpreter` pass would create an error for his test. I would argue that the lowering test in the added lit test file already serves this ordering test purpose.

Reply:
> The comment is what made me ask the question.
> 
> > ```
> > # If there already is a apply_registered_pass, insert after the existing one
> > ```
> 
> To me this reads as insert after the first one. If there are two or more, then it will still insert after the first one.

I will reword it.

Reply:
This comment has been reworded to reflect that the insertion point is always after the last pass in the current pass sequence.